man/fstrm_replay.1
.TH fstrm_replay 1

.SH NAME

fstrm_replay \- Replay saved Frame Streams data to a socket connection.

.SH SYNOPSIS

.B fstrm_replay -t \fIcontent-type\fB -r \fIfile\fB [ -r \fIfile\fB ... ]
.br
.B "	[ -u \fIsocket-path\fB ] [ -a \fIIP\fB -p \fIport\fB ]"

.PP

.B fstrm_replay --type \fIcontent-type\fB
.br
.B "	--read \fIfile\fB [ --read \fIfile\fB ... ]
.br
.B "	[ --unix \fIsocket-path\fB ] [ --tcp \fIIP\fB --port \fIport\fB ]"

.SH DESCRIPTION

.B fstrm_replay
connects to a Frame Streams receiver on either the given UNIX domain
\fIsocket-path\fR or TCP \fIaddress\fR and \fIport\fR, then reads
and sends data of the supplied \fIcontent-type\fR from the given
\fIfile(s)\fR.

.SH OPTIONS

.TP
.B -t \fIcontent-type\fB | --type \fIcontent-type\fB
Specify the \fIcontent-type\fR to read from files and send to the
Frame Streams socket. Files whose content-type differ are skipped.
If the socket server does not accept the \fIcontent-type\fR,
.B fstrm_replay
will fail.

.TP
.B -r \fIfile\fB | --read \fIfile\fB
Read data from \fIfile\fR. Multiple files can be given with multiple
\fB-r\fR options. Files which cannot be opened, or do not contain valid
Frame Streams data will be skipped.

.TP
.B -u \fIsocket-path\fB | --unix \fIsocket-path\fR
Connect to Unix domain \fIsocket-path\fR to write Frame Streams data.
Only one of \fB-u\fR or \fB-a\fR may be given.

.TP
.B -a \fIIP\fB | --tcp \fIIP\fB
Connect to TCP address \fIIP\fR to write Frame Streams data. Only one of
\fB-u\fR or \fB-a\fR may be given. Use of \fB-a\fR requires a port
given with \fB-p\fR.

.TP
.B -p \fIport\fB | --port \fIport\fB
If \fB-a\fR is given, use TCP port \fIport\fR to write Frame Streams
data.

.SH EXAMPLES

Replay dnstap data over UNIX domain socket:

.nf
	fstrm_replay -t protobuf:dnstap.Dnstap \\
		-u /var/run/named/dnstap.sock -r dnstap-log.fstrm
.fi

.SH SEE ALSO

.BR fstrm_capture (1),
.BR fstrm_dump (1),
.br
Frame Streams C Library \fBhttps://farsightsec.github.io/fstrm\fR


ChangeLog
fstrm (0.4.0)

  * Added manual pages for fstrm_capture and fstrm_dump.

  * Added new tool, fstrm_replay, for replaying saved Frame Streams data
    to a socket connection.

  * Adds TCP support. Add tcp_writer to the core library which
    implements a bi-directional Frame Streams writer as a TCP socket
    client. Introduces new developer API: fstrm_tcp_writer_init,
    fstrm_tcp_writer_options_init, fstrm_tcp_writer_options_destroy,
    fstrm_tcp_writer_options_set_socket_address, and
    fstrm_tcp_writer_options_set_socket_port.

  * fstrm_capture: new options for reading from TCP socket.

  * fstrm_capture: add "-c" / "--connections" option to limit the
    number of concurrent connections it will accept.

  * fstrm_capture: add "-b / --buffer-size" option to set the read buffer
    size (effectively the maximum frame size) to a value other than the
    default 256 KiB.

  * fstrm_capture: skip oversize messages to fix stalled connections
    caused by messages larger than the read highwater mark of the input
    buffer. Discarded messages are logged for the purposes of tuning
    the input buffer size.

  * fstrm_capture: complete sending of FINISH frame before closing
    connection.

  * Various test additions and improvements.

fstrm (0.3.2)

  * Accommodate systems without pthread_condattr_setclock (Issue #34)

 -- Chris Mikkelson <cmikk@fsi.io>  Tue, 21 Mar 2017 19:15:11 -0500

fstrm (0.3.1)

  * Add support for '-' as a filename for stdin/stdout (PR #28)

  * destroy condition variable and mutexes in fstrm_iothr_destroy() (PR #25)

 -- Chris Mikkelson <cmikk@fsi.io>  Thu, 26 Jan 2017 16:05:06 -0600

fstrm (0.3.0)

  * Output file rotation (Issue #22).

  * Build on OS X (Issue #21).

  * Change state properly in fstrm_writer_close() (Issue #18).

  * force output flush on sighup (Issue #12).
 
 -- Ben April <bapril@fsi.io>  Fri, 24 Jun 2016 22:29:56 -0500

fstrm (0.2.0)

  * Fix fstrm_writer_open() to allow multiple opens (Issue #1).

  * New utility 'fstrm_capture', which requires libevent >= 2 (Issue #4).

  * Fix incorrect pthread_cond_timedwait() usage (Issue #5).

  * Backwards incompatible library API and protocol changes.

  * Doxygen documentation.

 -- Robert Edmonds <edmonds@fsi.io>  Thu, 06 Nov 2014 19:19:44 -0500

fstrm (0.1.0)

  * Initial release.

 -- Robert Edmonds <edmonds@fsi.io>  Tue, 11 Feb 2014 20:00:25 -0500


src/fstrm_replay.c
/*
 * Copyright (c) 2018 by Farsight Security, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <arpa/inet.h>
#include <sys/uio.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#include <fstrm.h>

#include "libmy/argv.h"

struct replay {
	struct replay_args 	*args;
	uint8_t 		*content_type;
	size_t			len_content_type;
};

struct replay_args {
	bool			help;
	char			*content_type;
	char			*unix_address;
	char			*tcp_address;
	char			*tcp_port;
	argv_array_t		files;
};

static struct replay		g_program_ctx;
static struct replay_args	g_program_args;

static argv_t g_args[] = {
	{ 'h',  "help",
		ARGV_BOOL,
		&g_program_args.help,
		NULL,
		"display this help text and exit" },

	{ 't',  "type",
		ARGV_CHAR_P | ARGV_FLAG_MAND,
		&g_program_args.content_type,
		"<STRING>",
		"Frame Streams content type" },

	{ 'u',  "unix",
		ARGV_CHAR_P,
		&g_program_args.unix_address,
		"<FILENAME>",
		"Unix socket path to write to" },

	/* ARGV_ONE_OF indicates that the user must specify the
	 * previous option (-u) or next option (-a), but not both.
	 */
	{ ARGV_ONE_OF },

	{ 'a',  "tcp",
		ARGV_CHAR_P,
		&g_program_args.tcp_address,
		"<ADDRESS>",
		"TCP socket address to write to" },

	{ 'p',  "port",
		ARGV_CHAR_P,
		&g_program_args.tcp_port,
		"<PORT>",
		"TCP socket port to write to" },

	{ 'r', "read-file",
		ARGV_CHAR_P | ARGV_FLAG_ARRAY | ARGV_FLAG_MAND,
		&g_program_args.files,
		"<FILE>",
		"Files to read Frame Streams data from" },

	{ ARGV_LAST },
};

static void
usage(const char *msg)
{
	if(msg)
		fprintf(stderr, "%s: Usage error: %s\n", argv_program, msg);
	argv_usage(g_args, ARGV_USAGE_DEFAULT);
	argv_cleanup(g_args);
	exit(EXIT_FAILURE);
}

static bool
parse_args(int argc, char **argv, struct replay *ctx)
{
	if (argv_process(g_args, argc, argv) != 0)
		return false;

	if (g_program_args.help)
		return false;

	if ((g_program_args.tcp_address != NULL) &&
		(g_program_args.tcp_port == NULL))
		usage("--tcp requires --port");

	ctx->content_type = (uint8_t *)g_program_args.content_type;
	ctx->len_content_type = strlen(g_program_args.content_type);
	return true;
}

static struct fstrm_writer *
init_writer(void)
{
	struct fstrm_writer *w;
	struct fstrm_writer_options *wopt;
	fstrm_res res;

	/* Setup writer options. */
	wopt = fstrm_writer_options_init();
	res = fstrm_writer_options_add_content_type(wopt,
				g_program_ctx.content_type,
				g_program_ctx.len_content_type);
	if (res != fstrm_res_success) {
		fstrm_writer_options_destroy(&wopt);
		return NULL;
	}

	if (g_program_args.unix_address != NULL) {
		struct fstrm_unix_writer_options *uwopt;

		if (g_program_args.tcp_port != NULL)
			fputs("Warning: Ignoring --port with --unix.\n", stderr);

		uwopt  = fstrm_unix_writer_options_init();
		fstrm_unix_writer_options_set_socket_path(uwopt, g_program_args.unix_address);
		w = fstrm_unix_writer_init(uwopt, wopt);
		fstrm_unix_writer_options_destroy(&uwopt);
		fstrm_writer_options_destroy(&wopt);
		if (w == NULL) {
			fputs("Error: fstrm_unix_writer_init() failed.\n", stderr);
			return NULL;
		}
	} else {
		struct fstrm_tcp_writer_options *twopt;
		unsigned long port;
		char *endptr;
		struct sockaddr_in sin;
		struct sockaddr_in6 sin6;

	        /* Parse TCP listen port. */
	        port = strtoul(g_program_args.tcp_port, &endptr, 0);
	        if (*endptr != '\0' || port > UINT16_MAX)
	                usage("Failed to parse TCP listen port");

		/* Parse TCP listen address. */
		if ((inet_pton(AF_INET, g_program_args.tcp_address, &sin) != 1) &&
		    (inet_pton(AF_INET6, g_program_args.tcp_address, &sin6) != 1))
			usage("Failed to parse TCP listen address");

		twopt = fstrm_tcp_writer_options_init();
		fstrm_tcp_writer_options_set_socket_address(twopt, g_program_args.tcp_address);
		fstrm_tcp_writer_options_set_socket_port(twopt, g_program_args.tcp_port);

		w = fstrm_tcp_writer_init(twopt, wopt);
		fstrm_tcp_writer_options_destroy(&twopt);
		fstrm_writer_options_destroy(&wopt);

		if (w == NULL) {
			fputs("Error: fstrm_tcp_writer_init() failed.\n", stderr);
			return NULL;
		}
	}

	res = fstrm_writer_open(w);
	if (res != fstrm_res_success) {
		fstrm_writer_destroy(&w);
		fputs("Error: fstrm_writer_open() failed.\n", stderr);
		return NULL;
	}

	return w;
}

static void
process_file(const char *fname, struct fstrm_writer *w)
{
	struct fstrm_reader *r = NULL;
	struct fstrm_file_options *fopt = NULL;
	const struct fstrm_control *control = NULL;
	fstrm_res res;

	fopt = fstrm_file_options_init();
	fstrm_file_options_set_file_path(fopt, fname);

	/* Initialize file reader. */
	r = fstrm_file_reader_init(fopt, NULL);
	fstrm_file_options_destroy(&fopt);
	if (r == NULL) {
		fprintf(stderr, "Warning: failed to open %s, skipping\n", fname);
		return;
	}

	/* Check file content type. */
	res = fstrm_reader_get_control(r, FSTRM_CONTROL_START, &control);
	if (res != fstrm_res_success) {
		fprintf(stderr, "Warning: failed to read control frame from %s, skipping\n", fname);
		fstrm_reader_destroy(&r);
		return;
	}

	res = fstrm_control_match_field_content_type(control,
				g_program_ctx.content_type,
				g_program_ctx.len_content_type);
	if (res != fstrm_res_success) {
		fprintf(stderr, "Warning: content type mismatch for %s, skipping\n", fname);
		fstrm_reader_destroy(&r);
		return;
	}

	/* Loop over file data. */
	for (;;) {
		const uint8_t *data;
		size_t len_data;

		res = fstrm_reader_read(r, &data, &len_data);
		if (res == fstrm_res_success) {
			/* Write the data frame. */
			res = fstrm_writer_write(w, data, len_data);
			if (res != fstrm_res_success) {
				fputs("Error: write_data_frame() failed", stderr);
				exit(EXIT_FAILURE);
			}
		} else if (res == fstrm_res_stop) {
			/* Normal end of data stream. */
			res = fstrm_reader_get_control(r, FSTRM_CONTROL_STOP, &control);
			if (res != fstrm_res_success) {
				fprintf(stderr, "Error: unable to read STOP frame from %s.\n", fname);
			}
			break;
		} else {
			/* Abnormal end. */
			fprintf(stderr, "Error: fstrm_reader_read() failed.\n");
			break;
		}
	}
	fstrm_reader_destroy(&r);
}

int main(int argc, char **argv)
{
	int i;
	fstrm_res res;
	struct fstrm_writer *w = NULL;

	if (!parse_args(argc, argv, &g_program_ctx))
		usage(NULL);

	w = init_writer();

	for (i = 0; i < ARGV_ARRAY_COUNT(g_program_args.files); i++) {
		const char *in_fname;
		in_fname = ARGV_ARRAY_ENTRY(g_program_args.files, char *, i);
		process_file(in_fname, w);
	}

	res = fstrm_writer_close(w);
	if (res != fstrm_res_success)
		fputs("Error: fstrm_writer_close() failed", stderr);

	fstrm_writer_destroy(&w);
	return 0;
}
